docs/plugins.md

```diff
@@ -1,5 +1,202 @@
 # Plugin System
 
-The plugin system will allow to manipulate requests and responses inside a `HttpClient`.
+The plugin system allow to manipulate requests and responses inside an `HttpClient`.
 
-TODO: finalize system and document.
+By using the `Http\Client\Plugin\PluginClient` you can inject an `HttpClient` or an `HttpAsyncClient`, and an array
+of plugins implementing the `Http\Client\Plugin\Plugin` interface.
+
+Each plugin can modify the `RequestInterface` sent or the `ResponseInterface` received. It can also change the behavior of a call
+like retrying the request or emit another one when a redirection is present.
+
+## Install
+
+Install the plugin client in your project with composer:
+
+``` bash
+composer require "php-http/plugins:^1.0"
+```
+
+## Usage
+
+First you need to have some plugins:
+
+```php
+use Http\Client\Plugin\RetryPlugin;
+use Http\Client\Plugin\RedirectPlugin;
+
+$retryPlugin = new RetryPlugin();
+$redirectPlugin = new RedirectPlugin();
+```
+
+Then you can create a `PluginClient`:
+
+```php
+use Http\Discovery\HttpClientDiscovery;
+use Http\Client\Plugin\PluginClient;
+
+...
+
+$pluginClient = new PluginClient(HttpClientDiscovery::find(), [
+    $retryPlugin,
+    $redirectPlugin
+]);
+```
+
+After you can use this plugin client like a classic `Http\Client\HttpClient` or `Http\Client\HttpAsyncClient` one:
+
+```php
+// Send a request
+$response = $pluginClient->sendRequest($request);
+
+// Send an asynchronous request
+$promise  = $pluginClient->sendAsyncRequest($request);
+```
+
+Go to the [tutorial](tutorial.md) to read more about using `HttpClient` and `HttpAsyncClient`
+
+## Available plugins
+
+Each plugin have its own configuration and dependencies, check the documentation for each of the available plugin:
+
+ - [Authentication](plugins/authentication.md): Add authentication header on a request
+ - [Cookie](plugins/cookie.md): Add cookie to request and save them from the response by using a CookieJar
+ - [Encoding](plugins/encoding.md): Add support for receiving chunked, deflate or gzip response 
+ - [Error](plugins/error.md): Transform bad response (400 to 599) to exception
+ - [Redirect](plugins/redirect.md): Follow redirection coming from 3XX responses
+ - [Retry](plugins/retry.md): Retry a failed call
+ - [Stopwatch](plugins/stopwatch.md): Log time of a request call by using the Stopwatch component
+
+## Order of plugins
+
+When you inject an array of plugins into the `PluginClient`, order of the plugins in array matters.
```

Comment:
I was given quite the opposite advice (as we use in french to have something similar that "the", and we told me, it was a bad habit to do so in english, but it may depend if it's written or spoken ?)